rsocket-core/src/main/java/io/rsocket/metadata/CompositeMetadataFlyweight.java

```diff
@@ -0,0 +1,492 @@
+   * result, the first byte can simply be skipped, and the remaining of the buffer be decoded to the
+   * mime type.
+   *
+   * <p>If the mime header buffer is less than 2 bytes long, returns {@code null}.
+   *
+   * @param flyweightMimeBuffer the mime header {@link ByteBuf} that contains length + custom mime
+   *     type
+   * @return the decoded custom mime type, as a {@link CharSequence}, or null if the input is
+   *     invalid
+   * @see #decodeMimeIdFromMimeBuffer(ByteBuf)
+   */
+  @Nullable
+  public static CharSequence decodeMimeTypeFromMimeBuffer(ByteBuf flyweightMimeBuffer) {
+    if (flyweightMimeBuffer.readableBytes() < 2) {
+      return null;
+    }
+    // the encoded length is assumed to be kept at the start of the buffer
+    // but also assumed to be irrelevant because the rest of the slice length
+    // actually already matches _decoded_length
+    flyweightMimeBuffer.skipBytes(1);
+    int mimeStringLength = flyweightMimeBuffer.readableBytes();
+    return flyweightMimeBuffer.readCharSequence(mimeStringLength, CharsetUtil.US_ASCII);
+  }
+
+  /**
+   * Encode a {@link WellKnownMimeType well known mime type} and a metadata value length into a
+   * newly allocated {@link ByteBuf}.
+   *
+   * <p>This compact representation encodes the mime type via its ID on a single byte, and the
+   * unsigned value length on 3 additional bytes.
+   *
+   * @param allocator the {@link ByteBufAllocator} to use to create the buffer.
+   * @param mimeType a byte identifier of a {@link WellKnownMimeType} to encode.
+   * @param metadataLength the metadata length to append to the buffer as an unsigned 24 bits
+   *     integer.
+   * @return the encoded mime and metadata length information
+   */
+  static ByteBuf encodeMetadataHeader(
+      ByteBufAllocator allocator, byte mimeType, int metadataLength) {
+    ByteBuf buffer = allocator.buffer(4, 4).writeByte(mimeType | STREAM_METADATA_KNOWN_MASK);
+
+    NumberUtils.encodeUnsignedMedium(buffer, metadataLength);
+
+    return buffer;
+  }
+
+  /**
+   * Encode a custom mime type and a metadata value length into a newly allocated {@link ByteBuf}.
+   *
+   * <p>This larger representation encodes the mime type representation's length on a single byte,
+   * then the representation itself, then the unsigned metadata value length on 3 additional bytes.
+   *
+   * @param allocator the {@link ByteBufAllocator} to use to create the buffer.
+   * @param customMime a custom mime type to encode.
+   * @param metadataLength the metadata length to append to the buffer as an unsigned 24 bits
+   *     integer.
+   * @return the encoded mime and metadata length information
+   */
+  static ByteBuf encodeMetadataHeader(
+      ByteBufAllocator allocator, String customMime, int metadataLength) {
+    ByteBuf mimeBuffer = allocator.buffer(customMime.length());
+    mimeBuffer.writeCharSequence(customMime, CharsetUtil.UTF_8);
```

Comment:
It's probably easier to use ByteBufUtil.writeAscii(allocator, customMime);  to create the mimeBuffer

Reply:
I pushed a change that simplifies that encode method, avoiding the use of a `CompositeByteBuf`. But I think UTF8 writing is still required to detect non-ASCII chars, as `writeAscii` silently removes such characters (see commit message)